ext/standard/assert.c

```diff
@@ -171,41 +172,34 @@ PHP_FUNCTION(assert)
 
 		ZVAL_FALSE(&retval);
 
-		/* XXX do we want to check for error here? */
-		if (!description) {
+		if (!description_str && !description_obj) {
 			call_user_function(NULL, NULL, &ASSERTG(callback), &retval, 3, args);
 			zval_ptr_dtor(&(args[2]));
 			zval_ptr_dtor(&(args[0]));
-		} else {
-			ZVAL_STR(&args[3], zval_get_string(description));
+		} else if (description_str) {
+			ZVAL_STR(&args[3], description_str);
 			call_user_function(NULL, NULL, &ASSERTG(callback), &retval, 4, args);
 			zval_ptr_dtor(&(args[3]));
 			zval_ptr_dtor(&(args[2]));
 			zval_ptr_dtor(&(args[0]));
+		} else {
+			zend_argument_type_error(2, "must be of type string for the ASSERT_CALLBACK mode");
 		}
 
 		zval_ptr_dtor(&retval);
 	}
 
 	if (ASSERTG(exception)) {
-		if (!description) {
-			zend_throw_exception(assertion_error_ce, NULL, E_ERROR);
-		} else if (Z_TYPE_P(description) == IS_OBJECT &&
-			instanceof_function(Z_OBJCE_P(description), zend_ce_throwable)) {
-			Z_ADDREF_P(description);
```

Comment:
Should I add this ref to description_obj? I haven't found a way to do so in case of `zend_object`s :(